nf_core/components/update.py

```diff
@@ -0,0 +1,851 @@
+
+        patch_path = Path(self.dir / patch_relpath)
+        component_relpath = Path(self.component_type, repo_path, component)
+
+        # Check that paths in patch file are updated
+        self.check_patch_paths(patch_path, component)
+
+        # Copy the installed files to a new temporary directory to save them for later use
+        temp_dir = Path(tempfile.mkdtemp())
+        temp_component_dir = temp_dir / component
+        shutil.copytree(component_install_dir, temp_component_dir)
+
+        try:
+            new_files = ModulesDiffer.try_apply_patch(component, repo_path, patch_path, temp_component_dir)
+        except LookupError:
+            # Patch failed. Save the patch file by moving to the install dir
+            shutil.move(patch_path, Path(component_install_dir, patch_path.relative_to(component_dir)))
+            log.warning(
+                f"Failed to apply patch for {self.component_type[:-1]} '{component_fullname}'. You will have to apply the patch manually"
+            )
+            return False
+
+        # Write the patched files to a temporary directory
+        log.debug("Writing patched files")
+        for file, new_content in new_files.items():
+            fn = temp_component_dir / file
+            with open(fn, "w") as fh:
+                fh.writelines(new_content)
+
+        # Create the new patch file
+        log.debug("Regenerating patch file")
+        ModulesDiffer.write_diff_file(
+            Path(temp_component_dir, patch_path.relative_to(component_dir)),
+            component,
+            repo_path,
+            component_install_dir,
+            temp_component_dir,
+            file_action="w",
+            for_git=False,
+            dsp_from_dir=component_relpath,
+            dsp_to_dir=component_relpath,
+        )
+
+        # Move the patched files to the install dir
+        log.debug("Overwriting installed files installed files  with patched files")
+        shutil.rmtree(component_install_dir)
+        shutil.copytree(temp_component_dir, component_install_dir)
+
+        # Add the patch file to the modules.json file
+        self.modules_json.add_patch_entry(
+            component, self.modules_repo.remote_url, repo_path, patch_relpath, write_file=write_file
+        )
+
+        return True
+
+    def get_modules_subworkflows_to_update(self, component, modules_repo):
+        """Get all modules and subworkflows linked to the updated component."""
+        mods_json = self.modules_json.get_modules_json()
+        modules_to_update = []
+        subworkflows_to_update = []
+        installed_by = mods_json["repos"][modules_repo.remote_url][self.component_type][modules_repo.repo_path][
+            component
+        ]["installed_by"]
+
+        if self.component_type == "modules":
+            # All subworkflow names in the installed_by section of a module are subworkflows using this module
+            # We need to update them too
```

Comment:
what if there is no updated subworkflow for the new module yet?

Reply:
There will always be something for that git-sha, even if the files don't change.